aten/src/ATen/native/cuda/Loops.cuh

```diff
@@ -294,9 +297,9 @@ __global__ void unrolled_elementwise_kernel_for_multi_outputs(int N, func_t f, a
 template <int num_outputs, typename func_t, typename array_t, typename inp_calc_t, typename out_calc_t>
 static inline void launch_unrolled_kernel_for_multi_outputs(int64_t N, const func_t& f, array_t data, inp_calc_t ic, out_calc_t oc) {
   TORCH_INTERNAL_ASSERT(N > 0 && N <= std::numeric_limits<int32_t>::max());
-  int64_t grid = (N + block_work_size() - 1) / block_work_size();
+  int64_t grid = (N + block_work_size_dynamic() - 1) / block_work_size_dynamic();
```

Comment:
this won't work, because kernel itself uses `block_work_size()` https://github.com/pytorch/pytorch/pull/69942/files#diff-f711d0591df05ecc86728505605bb36286db1b3a8ba9c4957306c1142acbef1eR293, so if `block_work_size_dynamic` and `block_work_size` diverge, kernel will start producing wrong results. 
Changes like this are especially worrying because we don't have tests to make sure they are safe (or errors will manifest at some time in the future)

Reply:
Would it be acceptable to if def usage of block_work_size and block_work_size_dynamic on rocm vs cuda? We need to support launching kernels dynamically based on the hardware.

Reply:
But it would produce wrong results on ROCm also, unless you ifdef it in the kernel ifself, but you can't call `block_work_size_dynamic` in the kernel because it in turn calls getDeviceProperties.

Reply:
Are there any problems for ROCm if we constexpr `num_threads()` as 128, regardless of warp size?

Reply:
We have tested this with our internal qa and we are getting the behavior that we expect on the hardware platforms that we care about. The C10_WARP_SIZE reports the right warpsize when used on device code, our problems arise when we try to use that macro on the host side. That is why we have to use at::cuda::warp_size(). The at::cuda::warp_size() reports the right hardware warpsize.

Reply:
I think that will have a performance impact on our hardware with warpsize of 64. which would have launched with 256 threads.

Reply:
4 warps that cuda is launching don't have any sacred meaning, it was just a result of some rough benchmarking, and perf between 128 threads and 256 on cuda is roughly similar. I think we are overthinking a relatively simple problem here, unless you guys have data that you really need 128 threads on some models and 256 on others, depending on the warp size. This might be the case for some more complicated reduction kernels, but pointwise kernels are usually pretty tolerant to the block size.

Reply:
So this pr is a piece of the bigger pr here, https://github.com/pytorch/pytorch/pull/67294/files, which affects a lot of kernels. I am not sure all of them are pointwise kernels. We will do a some benchmarking and get back to you.

Reply:
> this won't work, because kernel itself uses `block_work_size()` https://github.com/pytorch/pytorch/pull/69942/files#diff-f711d0591df05ecc86728505605bb36286db1b3a8ba9c4957306c1142acbef1eR293, so if `block_work_size_dynamic` and `block_work_size` diverge, kernel will start producing wrong results. Changes like this are especially worrying because we don't have tests to make sure they are safe (or errors will manifest at some time in the future)

I agree that if `block_work_size_dynamic` and `block_work_size` diverge the kernel produces wrong results.  However, they should not diverge here.  For ROCm, each device (gfx target) has its own compilation pass and produces its own code object.  For gfx900 the `block_work_size` will be the constexpr which derives from `C10_WARP_SIZE` aka `warpSize`.  ROCm defines the `warpSize` constant at compile-time based on the gfx target.  This results in separate code objects for gfx900 (warpSize 64) versus gfx1030 (warpSize 32).  At runtime, the `block_work_size_dynamic` will be either 32 or 64 depending on the result of the device query.  Further, the correct corresponding code object is loaded for the gfx target.  That results in the compile-time and run-time warpSize agreeing.

Reply:
@ngimel I have been working on trying to use a constant number of threads(256) and I am getting mismatch results. I have fixed some of them but others are proving harder. It would involve altering multiple kernels in which I have no domain expertise in.  Further more, fixing the number of threads would be short a term solution that will fall apart when things change. Does the explanation provided by @jeffdaily ease your concerns when it comes to mismatch issues with a dynamic warp size look up? Is there anything that we can do so that this pr with dynamic warp size lookup is accepted?

Reply:
@ngimel @malfet Hope you had a good vacation. Any updates on this?

Reply:
We switched to a constant number.